
---------

# 📊 **`Variability, Dispersion, and Spread of Data in Statistics for Data Science`** 📈  

## 🔹 **What is Variability (Dispersion/Spread)?**  

✅ **Definition**: Variability (also called **dispersion or spread**) measures how **spread out** or **scattered** the data points are in a dataset.  
✅ **Why is it Important?**  
✔ Helps **understand data distribution** 📊.  
✔ Identifies **consistency vs. fluctuations** in data.  
✔ Critical for **machine learning models** to detect bias & variance.  

📌 **Key Metrics of Variability**:  
1️⃣ **Range** → Difference between max & min values.  
2️⃣ **Interquartile Range (IQR)** → Spread of the middle 50% of data.  
3️⃣ **Variance (σ², s²)** → Average squared deviation from the mean.  
4️⃣ **Standard Deviation (σ, s)** → Square root of variance, representing actual spread.  
5️⃣ **Coefficient of Variation (CV)** → Standardized measure of spread.  

---

## 🔹 **1. Range (Simplest Measure of Spread)📏**  

✅ **Definition**: The difference between **maximum and minimum** values.  
✅ **Formula**:  

$$
\text{Range} = X_{\text{max}} - X_{\text{min}}
$$

📌 **Example Calculation**  
📌 Data: **[10, 20, 30, 40, 50]**  
$$
\text{Range} = 50 - 10 = 40
$$

📌 **Python Code for Range**  



In [1]:

data = [10, 20, 30, 40, 50]
data_range = max(data) - min(data)
print("Range:", data_range)


Range: 40



📊 **Output**: `Range: 40`  

✅ **When to Use?**  
✔ Quick estimate of data spread.  
❌ **Sensitive to outliers** (not a reliable measure).  


---

## 🔹 **2. Interquartile Range (IQR) 🏗️**  

✅ **Definition**: The range of the **middle 50% of data** (from Q1 to Q3).  
✅ **Formula**:  
$$
\text{IQR} = Q_3 - Q_1
$$
where:  
- **Q1** = 25th percentile (lower quartile)  
- **Q3** = 75th percentile (upper quartile)  

📌 **Example Calculation**  
📌 Data: **[10, 20, 30, 40, 50]**  
✔ **Q1 = 20**, **Q3 = 40**  
✔ **IQR = 40 - 20 = 20**  

📌 **Python Code for IQR Calculation**  


In [2]:

import numpy as np

data = [10, 20, 30, 40, 50]
Q1 = np.percentile(data, 25)
Q3 = np.percentile(data, 75)
IQR = Q3 - Q1
print("Interquartile Range (IQR):", IQR)


Interquartile Range (IQR): 20.0




📊 **Output**: `IQR: 20`  

✅ **When to Use?**  
✔ **Best for skewed data** (not affected by extreme outliers).  
✔ **Used in box plots** for detecting outliers.  




---

## 🔹 **3. Variance (σ², s²) 🔄**  

✅ **Definition**: Measures the **average squared deviation from the mean**.  
✅ **Formula**:  
$$
\text{Variance} = \frac{\sum (X_i - \bar{X})^2}{N} \quad \text{(for population)}
$$

$$
\text{Variance} = \frac{\sum (X_i - \bar{X})^2}{n - 1} \quad \text{(for sample)}
$$

where: 

- $X_i$ = Data points  
- $\bar{X}$ = Mean of dataset  
- $N$, $n$ = Population or sample size

📌 **Example Calculation**  
📌 Data: **[10, 20, 30, 40, 50]**  
✔ Mean $\bar{X} = 30$
✔ Variance = **200**

📌 **Python Code for Variance Calculation**  



In [4]:

import numpy as np

data = [10, 20, 30, 40, 50]
population_variance = np.var(data)  # Population variance
sample_variance = np.var(data, ddof=1)  # Sample variance
print("\nPopulation Variance:", population_variance)
print("\nSample Variance:", sample_variance)



Population Variance: 200.0

Sample Variance: 250.0



📊 **Output**:  
- `Population Variance: 200`  
- `Sample Variance: 250`  

✅ **When to Use?**  
✔ Useful for **statistical tests** (e.g., ANOVA).  
❌ **Not in real-world interpretations** (because units are squared).  
 


------



## 🔹 **4. Standard Deviation (σ, s) 📉**  

✅ **Definition**: The **square root of variance**, representing actual spread in the same units as the data.  
✅ **Formula**:  

$$
\text{Standard Deviation} = \sqrt{\text{Variance}}
$$

📌 **Example Calculation**  
📌 Data: **[10, 20, 30, 40, 50]**  
✔ Population Variance = **200**  
✔ Standard Deviation = $$\sqrt{200} = 14.14$$

📌 **Python Code for Standard Deviation** 

In [5]:

import numpy as np

data = [10, 20, 30, 40, 50]  

population_std = np.std(data)  # Population standard deviation
sample_std = np.std(data, ddof=1)  # Sample standard deviation
print("\nPopulation Standard Deviation:", population_std)
print("\nSample Standard Deviation:", sample_std)



Population Standard Deviation: 14.142135623730951

Sample Standard Deviation: 15.811388300841896



📊 **Output**:  
- `Population Standard Deviation: 14.14`  
- `Sample Standard Deviation: 15.81`  

✅ **When to Use?**  
✔ **Best measure of spread** (in the same units as data).  
✔ Used in **data normalization & ML preprocessing**.

----

## 🔹 **5. Coefficient of Variation (CV) 📊**  

✅ **Definition**: Standardized measure of variability (relative spread).  
✅ **Formula**:  

$$
\text{CV} = \frac{\sigma}{\mu} \times 100\%
$$

where:  
- $\sigma$ = Standard deviation  
- $\mu$ = Mean  

📌 **Example Calculation**  
📌 Data: **[10, 20, 30, 40, 50]**  
✔ Mean = **30**, Standard Deviation = **14.14**  
✔ **CV = (14.14 / 30) × 100 = 47.13%**  

📌 **Python Code for CV Calculation**  


In [6]:

import numpy as np

cv = (population_std / np.mean(data)) * 100
print("\nCoefficient of Variation (CV):", cv, "%")



Coefficient of Variation (CV): 47.14045207910317 %



📊 **Output**: `CV: 47.13%`  



✅ **When to Use?**  
✔ **Comparing spread across datasets** with different units.  
✔ Used in **finance (risk analysis)** and **data quality checks**.  



-------

## 🔹 **6. Summary Table of Measures of Variability**  

| **Measure** | **Formula** | **Best For** | **Affected by Outliers?** |
|------------|------------|-------------|-----------------|
| **Range** 📏 | $$ X_{\text{max}} - X_{\text{min}} $$ | Quick spread estimate | ✅ Yes |
| **IQR** 🏗️ | $$ Q3 - Q1 $$ | Skewed data, outlier detection | ❌ No |
| **Variance** 🔄 | $$ \frac{\sum (X_i - \bar{X})^2}{N} $$ | Statistical models | ✅ Yes |
| **Standard Deviation** 📉 | $$ \sqrt{\text{Variance}} $$ | Feature scaling, ML models | ✅ Yes |
| **Coefficient of Variation** 📊 | $$ \frac{\sigma}{\mu} \times 100\% $$ | Comparing datasets | ✅ Yes |


---

# 🎯 **Final Takeaways**  

✔ **Variability (dispersion/spread) tells us how much data deviates from the mean**.  
✔ **Standard Deviation is the most widely used metric**.  
✔ **IQR is best for skewed data & outlier detection**.  
✔ **CV is used to compare spread across different datasets**.  

---------